src/analysis/validation.jl

```diff
@@ -2,102 +2,238 @@
 """
 Perform EPANET hydraulic simulation (via WNTR) and compute timeseries of flows and heads
 """
-function validate(data::Dict{String,Any}, solution::Dict{String,Any},
-                     inpfilepath::String, wn, wnres)
 
-    # for each node, compute the head values
-    # for each link, compute the flow rates and head gains
-    
-    wslinks = wnres.link
-    wsnodes = wnres.node
-
-    # ------------------------------------------#
-    #                   node                    #
-    # ------------------------------------------#
-    wntr_head = wsnodes["head"]
-
-    node_head_wntr = Dict{String,Any}()
-    node_head_wm = Dict{String,Any}()
-
-    for (key, node) in solution["1"]["node"] 
-        node_name = data["node"][key]["source_id"][2]
-        node_head_wntr[node_name] = Array{Float64,1}(undef,length(solution))
-        node_head_wm[node_name] = Array{Float64,1}(undef,length(solution))
-    end
+using DataFrames
 
-    for tx in 1:length(keys(solution))
-        for (key, node) in solution[string(tx)]["node"] 
-            node_name = data["node"][key]["source_id"][2]
-            if haskey(wntr_head, node_name)
-                node_head_wntr[node_name][tx] = wntr_head[node_name].values[tx]
-                node_head_wm[node_name][tx] = node["h"]
-            end
-        end
-    end
+# compare nodes
+function get_node_dataframe(wm_data,wm_solution,wntr_data,wntr_simulation, node_id)
+    duration = length(wm_solution["solution"]["nw"])            # number of time steps
+    time_step = wm_data["time_step"]/3600                       # length per time step (hour)
```

Comment:
There are other portions of the code (in `simulation.jl`) where you assume that the time step to be used in the EPANET simulation (particularly when adding controls) is one hour. You should use the same logic that you're using here to compute the actual times at which the controls should be set. For example, I've seen examples where the time step is 1800 seconds and not one hour.

Reply:
Looks like Yang has done this.